runtime/lua/vim/lsp.lua

```diff
@@ -940,18 +940,20 @@ function lsp.start_client(config)
   --- This is a thin wrapper around {client.rpc.request} with some additional
   --- checks for capabilities and handler availability.
   ---
-  ---@param method (string) LSP method name.
-  ---@param params (table) LSP request params.
-  ---@param handler (function, optional) Response |lsp-handler| for this method.
-  ---@param bufnr (number) Buffer handle (0 for current).
+  ---@param method string LSP method name.
+  ---@param params table LSP request params.
+  ---@param handler function|nil Response |lsp-handler| for this method.
+  ---@param bufnr number Buffer handle (0 for current).
+  ---@param on_complete function|nil Callback function to be called after completion of the
+  -- request.
   ---@returns ({status}, [request_id]): {status} is a bool indicating
   ---whether the request was successful. If it is `false`, then it will
   ---always be `false` (the client has shutdown). If it was
   ---successful, then it will return {request_id} as the
   ---second result. You can use this with `client.cancel_request(request_id)`
   ---to cancel the-request.
   ---@see |vim.lsp.buf_request()|
-  function client.request(method, params, handler, bufnr)
+  function client.request(method, params, handler, bufnr, on_complete)
```

Comment:
Why do we need this in addition to the handler ? Isn't it possible to just the logic into the handler?

Reply:
No, it's not. We need to set the buffer modifiable in the event that the request is cancelled. If a user cancels a pending request the handler will never be called and the buffer will remain unmodifiable.